capsules/src/dals/README.md

```diff
@@ -0,0 +1,100 @@
+Dynamic App-Loading System (DALS)
+=================================
+
+## Description
+
+This document introduces the design for a novel system named **DALS** which allows 
+the Kernel to dynamically (at runtime) load and run apps.  The data for a new app 
+to be loaded and run by Tock is transmitted to a Tock-compatible board through any 
+standard  communication protocol (BLE, Zigbee, USB, etc.) which provides "acks" or a similar service
+to lower data rates.  This data is then pushed 
+segment-by-segment through **DALS** which places it in an empty space in flash space and
+then sets up the process-specific elements for the newly-loaded app.
+
+## Design
+
+The design of **DALS** consists of a series of capsules with one capsule, the **AppLoader**, considered
+to be trusted.  This main trusted capsule is the heart of **DALS** and contains the code which collects
```

Comment:
Several months ago, on the core WG call, we discussed the "capsule" terminology and whether or not a capsule can be trusted ([notes](https://github.com/tock/tock/blob/master/doc/wg/core/notes/core-notes-2020-04-03.md)). We came to the conclusion that "capsule" never refers to code that the core kernel trusts. You are correct to state that some component of the DALS [must be trusted by the core kernel](https://github.com/tock/tock/blob/master/doc/threat_model/Application_Loader.md).

Based on those discussions, I think the trusted code should live outside the "capsules" directory. The rest of DALS -- the code that doesn't need to be trusted under Tock's threat model -- can be in capsules.

Reply:
Hmmm... @bradjc would you agree with moving the "trusted code" outside of the capsules directory?

Reply:
Yes this is a good point. Something needs to check that the start of the app header is valid when it is written to flash. One way to do this would be to embed that check in the implementation of the flash interface given to DALS. Something (very roughly) like:

```rust
impl NonvolatileStorage for AppFlashWriter {
    fn write(offset, buffer, len) {
        if offset < 8 {
            // make sure header in buffer is valid
        } else {...}
    }
    ...
}
```

That implementation would 1) confine the app to loading only in the flash space allocated for the incoming app, and 2) verify that the start of the TBF header it writes to flash has the correct version and length.

The code outlined in this PR would then be untrusted, and the trust lies in a proper sandboxed nonvolatile storage interface being provided to DALS.